x/bank/keeper/grpc_query.go

```diff
@@ -113,3 +113,48 @@ func (k BaseKeeper) Params(ctx context.Context, req *types.QueryParamsRequest) (
 
 	return &types.QueryParamsResponse{Params: params}, nil
 }
+
+// DenomsMetadata implements Query/DenomsMetadata gRPC method.
+func (k BaseKeeper) DenomsMetadata(c context.Context, req *types.QueryDenomsMetadataRequest) (*types.QueryDenomsMetadataResponse, error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.DenomMetadataPrefix)
+
+	metadatas := []types.Metadata{}
+	pageRes, err := query.Paginate(store, req.Pagination, func(_, value []byte) error {
+		var metadata types.Metadata
+		k.cdc.MustUnmarshalBinaryBare(value, &metadata)
+
+		metadatas = append(metadatas, metadata)
+		return nil
+	})
+
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &types.QueryDenomsMetadataResponse{
+		Metadatas:  metadatas,
+		Pagination: pageRes,
+	}, nil
+}
+
+// DenomMetadata implements Query/DenomMetadata gRPC method.
+func (k BaseKeeper) DenomMetadata(c context.Context, req *types.QueryDenomMetadataRequest) (*types.QueryDenomMetadataResponse, error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	if req.Denom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid denom")
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	return &types.QueryDenomMetadataResponse{
+		Metadata: k.GetDenomMetaData(ctx, req.Denom),
```

Comment:
So if the denom doesn't exist, we return an empty metadata and no error? Is this the expected behaviour?

Reply:
I think we should return an error:
```
status.Errorf(codes.NotFound, "Denom %s not found", req.Denom)
```

Check `/x/staking/keeper/grpc_query.go` for example.

Reply:
It looks like the `GetDenomMetaData` panics on a non existing denom.